From 4e0ba4499f7eea9ce62ae0f7df5894da374bcad6 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:45 -0400
Subject: [fs] nfsd: Prepare nfsd4_close() for open stateid referencing

Message-id: <1410898849-22126-102-git-send-email-bfields@redhat.com>
Patchwork-id: 95734
O-Subject: [Patch RHEL7 101/165] nfsd: Prepare nfsd4_close() for open stateid referencing
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Prepare nfsd4_close for a future where nfs4_preprocess_seqid_op()
hands it a fully referenced open stateid. Necessary step toward
client_mutex removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 8a0b589d8fd0e63579982cbfda099a2e09b52811
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index f3018cb..4e50f14 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -4579,10 +4579,15 @@ nfsd4_close(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
  nfsd4_bump_seqid(cstate, status);
  if (status)
   goto out; 
+ /* FIXME: move into nfs4_preprocess_seqid_op */
+ atomic_inc(&stp->st_stid.sc_count);
  update_stateid(&stp->st_stid.sc_stateid);
  memcpy(&close->cl_stateid, &stp->st_stid.sc_stateid, sizeof(stateid_t));
 
  nfsd4_close_open_stateid(stp);
+
+ /* put reference from nfs4_preprocess_seqid_op */
+ nfs4_put_stid(&stp->st_stid);
 out:
  if (!cstate->replay_owner)
   nfs4_unlock_state();
-- 
1.7.1